Expand project guidelines (#8314)
ComplianceAsCode/content: CONTRIBUTING.md

Avoid merge commits in pull requests - use `git rebase` to get rid of them. This helps us keep git
tree clean and helps to make it easy to understand historical changes done in past commits.

> ### Maintainer info

Reviewer: ```suggestion
### Maintainer Information
```

Reviewers should pay attention to the following points:

The PR ...
* ... has been approved by an SME of the area that the PR addresses, which indicates that it is generally a step in a right direction.
* ... passes PR gating tests, or failing tests are waived by means of explaining the reason.
* ... adheres to the [coding style](docs/manual/developer/04_style_guide.md#complianceascode-style-guide).

> * ... has been tested. There are following options that allow for testing various aspects of the project:

Reviewer: ```suggestion
* ... has been tested. The following options that allow for testing various aspects of the project:
```

> * ... has been tested. The following options that allow for testing various aspects of the project:

Reviewer 2: nit: remove `that`

>   * [SSG Test Suite](tests/README.md) for rule tests that ensures that test OVAL checks and that test the ability of remediations to satisfy those checks.

Reviewer: ```suggestion
  * [SSG Test Suite](tests/README.md) for rule tests to ensure that OVAL checks are correct and test the ability of remediations to satisfy those checks.
```

>   * [SSG Test Suite](tests/README.md) for rule tests that ensures that test OVAL checks and that test the ability of remediations to satisfty those checks.

Reviewer 3: ```suggestion
  * [SSG Test Suite](tests/README.md) for rule tests that ensures that test OVAL checks and that test the ability of remediations to satisfy those checks.
```

  * [BATS framework](tests/unit/bash) for bash tests that allows for fast and exhaustive testing of remediations that are parametrized by Jinja2.
  * [Unit tests](tests/unit) that test components of the build system as well as components of the SSG Test Suite.

>   * Ad-hoc tests that are integrated into the `ctest` chain directly s.a. the shellcheck test.

Reviewer: ```suggestion
  * Ad-hoc tests that are integrated into the `ctest` chain directly, i.e. the shellcheck test.
```

* ... updates READMEs, man pages or other documentation when changes of described behavior are introduced.

> * ... doesn't contain merge commits - those can be removed by rebasing.

Reviewer 2: Might be helpful to link to a rebasing/git guide if we have one.

Author: We don't, please somebody suggest something that is both to the point and isn't located on some obscure blog that may go offline in a couple of years.

Reviewer 4: @matejak , this might be useful: https://github.com/marcusburghardt/ansible-role-openscap/blob/master/files/STARTGUIDE.md

It is included in the [ansible-role-openscap](https://galaxy.ansible.com/marcusburghardt/ansible_role_openscap).

Author: So for the time being, I would proceed with merging this PR as it is, because your guide doesn't contain all the info that one could be after, and supplying that would take some time. I see no issue with referencing it, or even with putting it here and referencing it from the roles project.

I think that following issues should be covered there:
- merge commits, how to recognize them
- what rebase does, and what can be accomplished by interactive rebase (reordering, merge of commits)
- a few words about conflict resolution - when to expect them and generally how to handle them

Reviewer: We can also turn on a setting in the project that forbids merge commits to help automate this requirement.

> * ... is rebased against recent-enough branch that executes the whole range of expected integrated gating tests.

Reviewer 2: nit: Maybe `relevant` versus `expected`? The expected behavior, IMO, would be for GitHub to auto-rebase the present target branch state and the proposed PR changes and then run CI on top of that, but sadly it doesn't and instead takes the base branch as-is. :-) Saying relevant would allow for some variation in main branch, but suggest that any tests which impact the proposed changes should be pulled in first. Thoughts?

* ... doesn't increase maintenance costs, and if it does, there is a substantial counterbalance.

> * ... doesn't impact the build time in a significant way.

Reviewer 2: Maybe `negatively impact`? :-) If we improve the build time significantly, that might be good to merge :D

Avoid merge commits in pull requests - use `git rebase` to get rid of them. This helps us keep git
tree clean and helps to make it easy to understand historical changes done in past commits.

### Maintainer Information

Reviewers should pay attention to the following points:

The PR ...
* ... has been approved by an SME of the area that the PR addresses, which indicates that it is generally a step in a right direction.
* ... passes PR gating tests, or failing tests are waived in a comment by the reviewer explaining the reasoning.
* ... adheres to the [coding style](docs/manual/developer/04_style_guide.md#complianceascode-style-guide).
* ... has been tested. The following options allow for testing various aspects of the project:
  * [SSG Test Suite](tests/README.md) for rule tests to ensure that OVAL checks are correct and test the ability of remediations to satisfy those checks.
    Every testable rule that is newly created or that just got modified has to have at least one test scenario.

>     If a PR interacts with such rule that has been previously without tests, the author as to supply a test scenario as part of PR to get it merged.

Reviewer: ```suggestion
    If a PR interacts with a testable without tests, the author shall supply a test scenario as part of the PR to get it merged.
```
